apps/emqx_mix_utils/lib/mix/tasks/emqx.xref.ex

```diff
@@ -0,0 +1,193 @@
+defmodule Mix.Tasks.Emqx.Xref do
+  use Mix.Task
+
+  alias Mix.Tasks.Emqx.Ct, as: ECt
+  alias EMQXUmbrella.MixProject, as: UMP
+
+  @requirements ["compile", "loadpaths"]
+
+  @xref :xref
+
+  @default_checks [
+    :undefined_function_calls,
+    :undefined_functions,
+    :locals_not_used,
+    :deprecated_function_calls,
+    # :warnings_as_errors,
+    :deprecated_functions,
+  ]
+
+  @excluded_apps MapSet.new([
+    :emqx_mix_utils,
+  ])
+
+  @exclusions MapSet.new([
+    EMQXUmbrella.MixProject,
+  ])
+
+  @queries [
+    {"E || \"mnesia\":\"dirty_delete.*\"/\".*\" : Fun", []},
+    {"E || \"mnesia\":\"transaction\"/\".*\" : Fun", []},
+    {"E || \"mnesia\":\"async_dirty\"/\".*\" : Fun", []},
+    {"E || \"mnesia\":\"clear_table\"/\".*\" : Fun", []},
+    {"E || \"mnesia\":\"create_table\"/\".*\" : Fun", []},
+    {"E || \"mnesia\":\"delete_table\"/\".*\" : Fun", []}
+  ]
+
+  @impl true
+  def run(args) do
+    start_xref()
+    set_library_path()
+    set_default()
+    add_umbrella_app_dirs()
+    check_failures = run_checks()
+    query_failures = run_queries()
+    stop_xref()
+    case {check_failures, query_failures} do
+      {[], []} ->
+        Mix.shell().info([
+          :green,
+          "No xref issues found"
+        ])
+        :ok
+
+      _ ->
+        Mix.shell().info([
+          :red,
+          "Xref encountered issues:\n\n",
+          inspect(%{
+            check_failures: check_failures,
+            query_failures: query_failures,
+          }, pretty: true),
+        ])
+    end
+
+    :ok
+  end
+
+  def run_checks() do
+    exclusions = gather_exclusions()
+
+    Enum.reduce(@default_checks, [], fn check, acc ->
+      {:ok, results} = :xref.analyse(@xref, check)
+
+      mods = Enum.flat_map(results, fn
+        {m, _f, _a} ->
+          [m]
+
+        {{ms, _fs, _as}, {_mt, _ft, _at}} ->
+          [ms]
+
+        _ ->
+          []
+      end)
+      mod_ignores =
+        mods
+        |> Enum.flat_map(&get_module_ignores/1)
+        |> MapSet.new()
+
+      ignores = MapSet.union(exclusions, mod_ignores)
+
+      ignored? = fn vertex ->
+        mod = case vertex do
+          {m, _, _} ->
+            m
+          _ ->
+            vertex
+        end
+        MapSet.member?(ignores, vertex) || MapSet.member?(ignores, mod)
+      end
+
+      results = Enum.reject(results, fn
+        {src, dest} ->
+          ignored?.(src) || ignored?.(dest)
+
+        vertex ->
+          ignored?.(vertex)
+      end)
+
+      if results == [] do
+        acc
+      else
+        [{check, results} | acc]
+      end
+    end)
+  end
+
+  def run_queries() do
+    Enum.flat_map(@queries, fn {query, expected} ->
+      {:ok, result} = :xref.q(@xref, to_charlist(query))
```

Comment:
Perhaps define them as charlist literals?

Reply:
In Elixir, it's more usual and convenient to both read and write strings (binaries), and charlists are mostly for interfacing with Erlang code (such as this).

IMO, since this is not a hot path nor it's likely to gain much in terms of speed, having the declaration being more convenient to read and write (without the `~c` sigils) is nicer.